sql: avoid unnecessary version bumps for UDTs during ALTER TABLE

spilchen opened the pull request:
Previously, in the legacy schema changer, we unconditionally updated backreferences from tables to user-defined types (UDTs) during ALTER TABLE operations. This happened even when the backreferences remained unchanged.

While this behavior was functionally harmless, it caused the version of the referenced types to be incremented unnecessarily.

This change avoids version bumps unless the backreference actually changes.

Fixes: #144293

Epic: none
Release note: none

fqazi reviewed (approved):
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144297)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)

___
*[`pkg/sql/catalog/typedesc/type_desc.go` line 426 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-ON_i0Te9qCA6xdcyCDJ:-ON_i0Te9qCA6xdcyCDK:baobdp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/37e9aa6f813df21dff583da685b30c255603644f/pkg/sql/catalog/typedesc/type_desc.go#L426)):*
> ```Go
> // RemoveReferencingDescriptorID removes the desired referencing descriptor ID
> // from the catalog.TypeDescriptor. It has no effect if the requested ID is not
> // present. A value of false is returned for that case.
> ```

Might be clearer to say it returns false if nothing is removed.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144297)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @spilchen)

___
*[`pkg/backup/restore_job.go` line 1404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-ON_iBHI1Agr1sMrShpq:-ON_iBHI1Agr1sMrShpr:b2htth7) ([raw file](https://github.com/cockroachdb/cockroach/blob/37e9aa6f813df21dff583da685b30c255603644f/pkg/backup/restore_job.go#L1404)):*
> ```Go
> 						return err
> 					}
> 					_ = typDesc.AddReferencingDescriptorID(table.GetID())
> ```

Nit: We could just modify these lines instead


<!-- Sent from Reviewable.io -->

rafiss commented:
nice find! due to https://github.com/cockroachdb/cockroach/pull/142921 and https://github.com/cockroachdb/cockroach/pull/140617, starting in v25.2, we will no longer fallback to the legacy schema changer when running ADD/DROP column on a regional by row table. so to get the most value out of this PR, we would need to backport it -- @spilchen and @fqazi: do you think a backport is worthwhile here?

fqazi commented:
Its a relatively low risk change, so probably worth doing it. It might also help reduce contention.

spilchen reviewed (commented):
Yeah, I can backport.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-:-ONaHSFMAeGeWTKPyvuj:b366oib)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @kev-cao)

___
*[`pkg/backup/restore_job.go` line 1404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-ON_iBHI1Agr1sMrShpq:-ONaGk7C1MxxogYQ9Hor:b-tlzl8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/37e9aa6f813df21dff583da685b30c255603644f/pkg/backup/restore_job.go#L1404)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: We could just modify these lines instead
</blockquote></details>

Are you suggesting we could skip this change? I can do that as it will reduce code churn. I only included it here because I wanted to be intentional about ignoring the return code.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144297)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @spilchen)

___
*[`pkg/backup/restore_job.go` line 1404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-ON_iBHI1Agr1sMrShpq:-ONaITmr8NvbECCs0pma:blrio7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/37e9aa6f813df21dff583da685b30c255603644f/pkg/backup/restore_job.go#L1404)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Are you suggesting we could skip this change? I can do that as it will reduce code churn. I only included it here because I wanted to be intentional about ignoring the return code.
</blockquote></details>

Done.

Yeah, but that makes sense. Its okay to leave this as explicit to avoid mystery.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: avoid unnecessary version bumps for UDTs during ALTER TABLE

Previously, in the legacy schema changer, we unconditionally updated
backreferences from tables to user-defined types (UDTs) during ALTER
TABLE operations. This happened even when the backreferences remained
unchanged.

While this behavior was functionally harmless, it caused the version of
the referenced types to be incremented unnecessarily.

This change avoids version bumps unless the backreference actually
changes.

Fixes: #144293

Epic: none
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144297)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @kev-cao)

___
*[`pkg/sql/catalog/typedesc/type_desc.go` line 426 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144297#-ON_i0Te9qCA6xdcyCDJ:-ONdczGx562wp4J5vWBI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37e9aa6f813df21dff583da685b30c255603644f/pkg/sql/catalog/typedesc/type_desc.go#L426)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Might be clearer to say it returns false if nothing is removed.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
